In [34]:
#Import pandas and numpy libraries to read and manipulate dataset
import pandas as pd
import numpy as np

In [35]:
#Read the file from desktop
df = pd.read_csv("C:/Users/praveena/Desktop/Bank_Project/dsm-bank-model-2870047/bankData/bank.csv")
df.head()

30   unemployed  married    primary  no  1787 no.1 no.2  cellular  19  oct  \
0  33     services  married  secondary  no  4789  yes  yes  cellular  11  may   
1  35   management   single   tertiary  no  1350  yes   no  cellular  16  apr   
2  30   management  married   tertiary  no  1476  yes  yes   unknown   3  jun   
3  59  blue-collar  married  secondary  no     0  yes   no   unknown   5  may   
4  35   management   single   tertiary  no   747   no   no  cellular  23  feb   

    79  1   -1  0  unknown no.3  
0  220  1  339  4  failure   no  
1  185  1  330  1  failure   no  
2  199  4   -1  0  unknown   no  
3  226  1   -1  0  unknown   no  
4  141  2  176  3  failure   no

In [36]:
#Dropping campaign related columns and recoginsing features and output columns
df1 = df.copy()
df1.drop(df1.iloc[:, 8:16], inplace = True, axis = 1)
df1.head()

30   unemployed  married    primary  no  1787 no.1 no.2 no.3
0  33     services  married  secondary  no  4789  yes  yes   no
1  35   management   single   tertiary  no  1350  yes   no   no
2  30   management  married   tertiary  no  1476  yes  yes   no
3  59  blue-collar  married  secondary  no     0  yes   no   no
4  35   management   single   tertiary  no   747   no   no   no

In [37]:
X = df.iloc[:, :-1].values               #Feature columns
y = df.iloc[:, -1].values                #Output variable

In [38]:
#Importing sklearn libraries for data pre-processing and splitting data to training and testing sets
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

In [39]:
df1.columns

Index(['30', 'unemployed', 'married', 'primary', 'no', '1787', 'no.1', 'no.2',
       'no.3'],
      dtype='object')

In [40]:
'''
   Recognising datatypes and naming columns to give processed data which is easy to model.
   First let's work on numerical columns and then categorical columns, from 'df' dataframe, 
   column numbers 0, 5 are numeric while others are categorical datatype.
'''
numeric_data = df1.iloc[:, [0, 5]].values
numeric_df = pd.DataFrame(numeric_data, dtype = object)
numeric_df.columns = ['age', 'balance']
numeric_df.head()

age balance
0  33    4789
1  35    1350
2  30    1476
3  59       0
4  35     747

In [41]:
X_categoric = df1.iloc[:, [1,2,3,4,6,7]].values

In [42]:
#applying StandardScaler to 'numeric_df' as both columns have values with a varied range
age_std_scale = StandardScaler()
numeric_df['age'] = age_std_scale.fit_transform(numeric_df[['age']])

In [43]:
balance_std_scale = StandardScaler()
numeric_df['balance'] = balance_std_scale.fit_transform(numeric_df[['balance']])
numeric_df.head()

age   balance
0 -0.772826  1.118549
1 -0.583699 -0.024115
2 -1.056517  0.017751
3  1.685824 -0.472675
4 -0.583699 -0.224472

In [44]:
#To apply ML models we need to convert categorical variables to numeric variables which can be done through "One-hot encoding".
one = OneHotEncoder()
categoric_data = one.fit_transform(X_categoric).toarray()
categoric_df = pd.DataFrame(categoric_data)
categoric_df.columns = one.get_feature_names_out()
categoric_df

x0_admin.  x0_blue-collar  x0_entrepreneur  x0_housemaid  x0_management  \
0           0.0             0.0              0.0           0.0            0.0   
1           0.0             0.0              0.0           0.0            1.0   
2           0.0             0.0              0.0           0.0            1.0   
3           0.0             1.0              0.0           0.0            0.0   
4           0.0             0.0              0.0           0.0            1.0   
...         ...             ...              ...           ...            ...   
4515        0.0             0.0              0.0           0.0            0.0   
4516        0.0             0.0              0.0           0.0            0.0   
4517        0.0             0.0              0.0           0.0            0.0   
4518        0.0             1.0              0.0           0.0            0.0   
4519        0.0             0.0              1.0           0.0            0.0   

      x0_retired  x0_self-employed  x0_services  x0_student  x0_technician  \
0            0.0               0.0          1.0         0.0            0.0   
1            0.0               0.0          0.0         0.0            0.0   
2            0.0               0.0          0.0         0.0            0.0   
3            0.0               0.0          0.0         0.0            0.0   
4            0.0               0.0          0.0         0.0            0.0   
...          ...               ...          ...         ...            ...   
4515         0.0               0.0          1.0         0.0            0.0   
4516         0.0               1.0          0.0         0.0            0.0   
4517         0.0               0.0          0.0         0.0            1.0   
4518         0.0               0.0          0.0         0.0            0.0   
4519         0.0               0.0          0.0         0.0            0.0   

      ...  x2_primary  x2_secondary  x2_tertiary  x2_unknown  x3_no  x3_yes  \
0     ...         0.0           1.0          0.0         0.0    1.0     0.0   
1     ...         0.0           0.0          1.0         0.0    1.0     0.0   
2     ...         0.0           0.0          1.0         0.0    1.0     0.0   
3     ...         0.0           1.0          0.0         0.0    1.0     0.0   
4     ...         0.0           0.0          1.0         0.0    1.0     0.0   
...   ...         ...           ...          ...         ...    ...     ...   
4515  ...         0.0           1.0          0.0         0.0    1.0     0.0   
4516  ...         0.0           0.0          1.0         0.0    0.0     1.0   
4517  ...         0.0           1.0          0.0         0.0    1.0     0.0   
4518  ...         0.0           1.0          0.0         0.0    1.0     0.0   
4519  ...         0.0           0.0          1.0         0.0    1.0     0.0   

      x4_no  x4_yes  x5_no  x5_yes  
0       0.0     1.0    0.0     1.0  
1       0.0     1.0    1.0     0.0  
2       0.0     1.0    0.0     1.0  
3       0.0     1.0    1.0     0.0  
4       1.0     0.0    1.0     0.0  
...     ...     ...    ...     ...  
4515    0.0     1.0    1.0     0.0  
4516    0.0     1.0    0.0     1.0  
4517    1.0     0.0    1.0     0.0  
4518    1.0     0.0    1.0     0.0  
4519    0.0     1.0    0.0     1.0  

[4520 rows x 25 columns]

In [45]:
#Now let's concatenate two dataframes
X_final = pd.concat([numeric_df, categoric_df], axis = 1)

In [15]:
#Using RandomForestClassifier as a ML model for data modeling
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators = 100)
rfc.fit(X_final, y)

RandomForestClassifier()

In [16]:
X_train, X_test, y_train, y_test=train_test_split(X_final,y, test_size= 0.2)
rfc = RandomForestClassifier(n_estimators = 100)
rfc.fit(X_train, y_train)

RandomForestClassifier()

In [17]:
y_pred=rfc.predict(X_test)
from sklearn.metrics import confusion_matrix
confusion_matrix=(y_test,y_pred)

In [18]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.8573008849557522

In [19]:
#changing estimator value to modify accuracy with same train n test split ratio
rfc1 = RandomForestClassifier(n_estimators = 50)
rfc1.fit(X_train, y_train)

RandomForestClassifier(n_estimators=50)

In [21]:
y_pred1=rfc1.predict(X_test)
confusion_matrix=(y_test,y_pred1)
accuracy_score(y_test,y_pred1)

0.8617256637168141

In [25]:
'''
    To improve model accuracy, considering only few main features and predicting the accuracy.
    We going to use only 0,1,2,3,5 columns as features to modify the accuracy of the model.
'''
X_categoric2 = df1.iloc[:, [1,2,3]].values
one = OneHotEncoder()
categoric_data = one.fit_transform(X_categoric2).toarray()
categoric_df2 = pd.DataFrame(categoric_data)
categoric_df2.columns = one.get_feature_names_out()
categoric_df2.columns

Index(['x0_admin.', 'x0_blue-collar', 'x0_entrepreneur', 'x0_housemaid',
       'x0_management', 'x0_retired', 'x0_self-employed', 'x0_services',
       'x0_student', 'x0_technician', 'x0_unemployed', 'x0_unknown',
       'x1_divorced', 'x1_married', 'x1_single', 'x2_primary', 'x2_secondary',
       'x2_tertiary', 'x2_unknown'],
      dtype='object')

In [27]:
X_final2 = pd.concat([numeric_df, categoric_df2], axis = 1)

In [30]:
rfc2 = RandomForestClassifier(n_estimators = 100)
rfc2.fit(X_final2, y)
X_train2, X_test2, y_train2, y_test2=train_test_split(X_final2,y, test_size= 0.2)
rfc2.fit(X_train2, y_train2)

RandomForestClassifier()

In [32]:
y_pred2=rfc2.predict(X_test2)
confusion_matrix=(y_test2,y_pred2)
accuracy_score(y_test2,y_pred2)

0.8816371681415929

Model accuracy improved from 0.857 to 0.88 when number of features are modified, though reducing decision trees of randomforest classifier did not help much.